Ensemble-Model.md

```diff
@@ -0,0 +1,1066 @@
+```
+
+
+```python
+training_masked = training.map(mask_addons).cache()
+```
+
+## Creating the feature matrices
+
+For each user, we want a matrix that contains a row for each whitelisted addon and a column for each recommender. A cell then contains the confidence score that the respective recommender gave for the respective user and addon.
+
+
+```python
+recommenders = {
+    "collaborative": CollaborativeRecommender(),
+    "similarity": SimilarityRecommender(),
+    "locale": LocaleRecommender("./top_addons_by_locale.json"),
+    "legacy": LegacyRecommender()
+}
+```
+
+    INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): s3-us-west-2.amazonaws.com
+    INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): s3-us-west-2.amazonaws.com
+    INFO:boto3.resources.action:Calling s3:get_object with {u'Bucket': 'telemetry-parquet', u'Key': 'taar/similarity/donors.json'}
+    INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): s3-us-west-2.amazonaws.com
+    INFO:boto3.resources.action:Calling s3:get_object with {u'Bucket': 'telemetry-parquet', u'Key': 'taar/similarity/lr_curves.json'}
+    INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): s3-us-west-2.amazonaws.com
+    INFO:boto3.resources.action:Calling s3:get_object with {u'Bucket': 'telemetry-parquet', u'Key': 'taar/locale/top10_dict.json'}
+    INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): s3-us-west-2.amazonaws.com
+    INFO:boto3.resources.action:Calling s3:get_object with {u'Bucket': 'telemetry-parquet', u'Key': 'taar/legacy/legacy_dict.json'}
+    INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): s3-us-west-2.amazonaws.com
+
+
+
+```python
+def compute_features(client_data):
+    recommendations = []
+    matrix = []
+    
+    for _, recommender in recommenders.items():
+        recommendations.append(recommender.get_weighted_recommendations(client_data))
+
+    for addon in whitelist:
+        matrix.append([features[addon] for features in recommendations])
+
+    return client_data, np.array(matrix)
+```
+
+
+```python
+X_unnormalized = training_masked.map(compute_features).cache()
+```
+
+    CPU times: user 1.3 s, sys: 20 ms, total: 1.32 s
+    Wall time: 1.34 s
+
+
+## Normalization
+
+The optimization algorithms that we use here are much more minimal than what you typically from highly optimized ML libs. Because of this, we need to take special care of properly preprocessing the data.
+
+In the following, we perform these operations:
+- [Min-Max scaling](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html)
+- Changing the locale scores to a double square root
+
+
+```python
+max_feature_values = X_unnormalized.map(lambda (_, features): np.max(features, axis=0)).reduce(np.maximum)
```

Comment:
> I.e. does a min_max transformation for the collaborative recommender mean the same thing as a min_max transformation for the similarity recommender?

Nope, unfortunately not. My plan was to find some transformations so that this is nearly the case, but I didn't get to that.